### ISSUE TITLE
Color corruption when downscaling PNG -> JPG

### ISSUE THREAD

**Issue by Draco18s** [libvips/libvips#3129]
**Describe the bug**
Haven't been able to identify the cause of the behavior, but reds get turned into oranges and blacks get darkened, turning dark gray gradients into black solid black abysses, while greens and blues are slightly boosted, leading to phantom whirls where there should be a smooth gradient:

Correct:
![image](https://user-images.githubusercontent.com/1750680/198928730-1ebffba1-d7e6-4e06-8844-b21e1c99de10.png)
Incorrect:
![image](https://user-images.githubusercontent.com/1750680/198928736-93da9bcc-dc3f-46f4-a87a-c0a189199ea8.png)

**To Reproduce**
1. Start with [this image](https://static1.e6ai.net/data/0a/1d/0a1d24533f0aafe2cf80e7b6411e9a86.png)
2. Pass it through [this function](https://github.com/zwagoth/e621ng/blob/2b8173e05a9a0d49f8d9ebbb0b8893c4425799a5/app/logical/danbooru_image_resizer.rb#L12-L18) with a maximum width of 850.
3. Compare results

**Expected behavior**
Expected: an 850 px wide downscaled version of the image with only minor differences in color due to jpg artifacting.

**Actual behavior**
[This image](https://static1.e6ai.net/data/sample/0a/1d/0a1d24533f0aafe2cf80e7b6411e9a86.jpg)
Green is leaking into the other channels in [very strange ways](https://user-images.githubusercontent.com/1750680/198998004-a5eb9803-2747-4813-82ec-6a3013107a21.png) (look at the red and magenta bars: there shouldn't be green there at all, but there it is, and it isn't even a linear gradient).

**Environment**
 - ruby-vips version 2.1.4
 - Ubuntu 20.04

**Comment by jcupitt** [comment]
Hi @Draco18s,

I've tried a few things and not been able to reproduce this. For example:

```
vipsthumbnail bad.png \
    --size "850<" \
    -o x.jpg[strip,interlace,optimize_coding,Q=90] \
    --no-rotate \
    --export-profile /usr/share/color/argyll/ref/sRGB.icm \
    --import-profile /usr/share/color/argyll/ref/sRGB.icm
```

Which I think should match your code gives me:

http://www.rollthepotato.net/~john/bad.png

http://www.rollthepotato.net/~john/x.jpg

Which look the same (try putting them in a pair of browser tabs and flipping the pages).

**Comment by jcupitt** [comment]
I tried with your updated image and that works fine for me too.

What version of libvips are you using? Is it the one that comes with ubuntu 20.04?

**Comment by Draco18s** [comment]
Hey thanks for poking around @jcupitt 

> What version of libvips are you using? Is it the one that comes with ubuntu 20.04?

ruby-vips version 2.1.4

> I tried with your updated image and that works fine for me too.

Yeah, someone else grabbed that from an LCD test image, which I hadn't thought to do.

We've had a couple ideas of what might be going wrong, from embedded color profiles not matching (ergo libvips applying the conversions twice), to PNG chunk or exif data causing issues, but every single test image we push through gets messed up in exactly the same way.

The weird part is that the same code is being used in another deployment and doesn't have the issue. At least, I've been told its the same code, there are front end and back end differences, but none of that should have touched the image rescaling portions.

(I should note that I don't have access to the code in question, I just happen to moderate a discord server and have contact with the folks that do manage the code as well the folks uploading images, so I'm being a liaison between several groups and took the initiative to file this report rather than wait on someone else to do it).

Working on getting some more information.

**Comment by jcupitt** [comment]
> ruby-vips version 2.1.4

ruby-vips is just a binding, it can work with many different libvips versions. It'll pick up the one it finds at runtime and present that API.

```
$ irb
irb(main):001:0> require 'vips'
=> true
irb(main):003:0> Vips::VERSION
=> "2.1.4"
irb(main):004:0> Vips::version_string
=> "8.14.0"
```

So that's ruby-vips 2.1.4 binding to libvips 8.14.0.

I think I'd try the libvips CLI interface and see if you can reproduce the issue there.

**Comment by r7vz9h3** [comment]
I managed to reproduce the bug with my short code, run on Debian on WSL2.

The color profile I used, from ArgyllCMS 2.0.0: [sRGB.zip](https://github.com/libvips/libvips/files/9905663/sRGB.zip)
(I got the same bad result with 2 other profiles I tried.)

```ruby
> Vips::VERSION
"2.1.4"
> Vips::version_string
"8.13.2-Sun Oct  2 18:10:46 UTC 2022"
```

The code (I didn't use ruby before today so it's messy):

```ruby
require "vips"

module DanbooruImageResizer
  extend self

  SRGB_PROFILE = "sRGB.icm" # Taken from ArgyllCMS 2.0.0
  # https://www.libvips.org/API/current/libvips-resample.html#vips-thumbnail
  THUMBNAIL_OPTIONS = { size: :down, linear: false, no_rotate: true, export_profile: SRGB_PROFILE, import_profile: SRGB_PROFILE }
  # https://www.libvips.org/API/current/VipsForeignSave.html#vips-jpegsave
  JPEG_OPTIONS = { background: 0, strip: true, interlace: true, optimize_coding: true }
  CROP_OPTIONS = { linear: false, no_rotate: true, export_profile: SRGB_PROFILE, import_profile: SRGB_PROFILE, crop: :attention }

  def resize(file, width, height, resize_quality = 90)
    output_file = Tempfile.new
    resized_image = thumbnail(file, width, height, THUMBNAIL_OPTIONS)
    resized_image.jpegsave(output_file.path, Q: resize_quality, **JPEG_OPTIONS)

    output_file
  end

  # https://github.com/libvips/libvips/wiki/HOWTO----Image-shrinking
  # https://www.libvips.org/API/current/Using-vipsthumbnail.md.html
  def thumbnail(file, width, height, options)
    Vips::Image.thumbnail(file.path, width, height: height, **options)
  rescue Vips::Error => e
    raise e unless e.message =~ /icc_transform/i
    Vips::Image.thumbnail(file.path, width, height: height, **options.except(:import_profile))
  end
end

im = File.open("color test image sRGB.png", "rb")

sample_im = DanbooruImageResizer.resize(im, 850, 671, 87)
# this creates a tmp file (for me in /tmp/) which is a jpg and shows the discoloration
```

Input image:

<img src="https://user-images.githubusercontent.com/116812064/199131050-0b4c40b3-bfdc-40aa-a105-0ae67e71d69b.png" width=50% height=50%>

Discolored output image, note the blue turning into purple, dark red into orange:

<img src="https://user-images.githubusercontent.com/116812064/199131936-5331f166-04cc-46fe-a8e7-0b2d8720d8b9.jpg" width=50% height=50%>

I will gladly provide more information if needed.

**Comment by r7vz9h3** [comment]
Even shorter code reproducing the behavior, this time in Python using `vips-8.13.1`:

```python
import pyvips

SRGB_PROFILE = 'sRGB.icm'
TEST_IMAGE = 'color test image sRGB.png'

JPEG_OPTIONS = {
    'background': 0,
    'strip': True,
    'interlace': True,
    'optimize_coding': True
}

THUMB_OPTIONS =  { 
    'size': 650,
    'linear': False,
    'no_rotate': True,
    'export_profile': SRGB_PROFILE,
    'import_profile': SRGB_PROFILE
}

resized = pyvips.Image.thumbnail(TEST_IMAGE, 800, height=600, **THUMB_OPTIONS)
resized.jpegsave('new_test_image.jpeg', Q=90, **JPEG_OPTIONS)
```

**Comment by Draco18s** [comment]
> > ruby-vips version 2.1.4
> 
> ruby-vips is just a binding, it can work with many different libvips versions. It'll pick up the one it finds at runtime and present that API.

Tells you what I know about Ruby. :P Just reporting what was reported to me. :)

**Comment by Maw-Fox** [comment]
via `conda` I can duplicate this on 8.13.1, on Ubuntu 20.04 (via `sudo`, 8.4.5) it is not an issue.

Strange dithering can be duplicated (on 8.13.1 only, afaik):
`vipsthumbnail test_image.png -s 600 -e sRGB.icm -i sRGB.icm -o tn_%s.jpeg[Q=90,interlace,optimize_coding,strip]`

This issue seems to come from whatever processing is happening with the `export-profile` option on both `resize` & `thumbnail` methods.

**Comment by evazion** [comment]
I noticed this after upgrading from 8.12.1 to 8.13.2. It appears to be caused by the `--export-profile` option in `vipsthumbnail`. Here's a minimal reproducer:

```bash
# The image from https://github.com/libvips/libvips/issues/3129#issuecomment-1297843356
curl https://user-images.githubusercontent.com/116812064/199131050-0b4c40b3-bfdc-40aa-a105-0ae67e71d69b.png -O in.png
vips thumbnail --export-profile srgb --size 2020x1594 in.png out.jpg
```

This works correctly in 8.12.1, but not in 8.13.2. It works when building vips from HEAD (17ca29adc9c52ef48cf5a9a59c26a83d0693fbbe). It also works when removing the `--export-profile srgb` option. So maybe this is the same bug that was fixed in #3027?

**Comment by jcupitt** [comment]
> Here's a minimal reproducer:

Ahh, I see this now, thanks!

Yes, this is in 8.13.2, but fixed 8.13.3 (HEAD of the 8.13 branch). I tried reverting the commit you linked and the error returned, so you're right, it's the same issue.

**Comment by jcupitt** [comment]
8.13 has picked up a few useful improvements now, let's push out 8.13.3 with this fix.

**Comment by jcupitt** [comment]
OK, we have 8.13.3:

https://github.com/libvips/libvips/releases/tag/v8.13.3

Thank you all for reporting and pushing on this! I'll close.

If there are still problems (hopefully not) please open a new issue.

**Comment by Draco18s** [comment]
Even though it turned out to be a known/fixed issue, looks like this was the impetus behind pushing a new version so that that fix is live.

My job here is done.

Thanks for your help!
